doc/termit.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH TERMIT 1 "NOV 2008" Linux "User Manuals"
.SH NAME
termit \(hy lightweight terminal emulator
.SH SYNOPSIS
.B termit [\-option ...]
.SH DESCRIPTION
.B termit
is a vte\(hybased lightweight terminal emulator. All configuration
is done via Lua\(hyscripts. The only other dependencies are 
Gtk+ and Vte.
.SH OPTIONS
.BR \-h,
.BR \-\-help
.RS
Print help
.RE
.BR \-v,
.BR \-\-version
.RS
Print version number
.RE
.BR \-e,
.BR \-\-execute
=
.I cmd
.RS
Use
.I cmd
as shell
.RE
.BR \-i,
.BR \-\-init
=
.I file
.RS
Use
.I file
instead of default rc.lua
.RE
.BR \-n,
.BR \-\-name
=
.I name
.RS
Use
.I name
in window name hint
.RE
.BR \-c,
.BR \-\-class
=
.I class
.RS
Use
.I class
in window class hint
.RE
.BR \-r,
.BR \-\-role
=
.I role
.RS
Use
.I role
in window role Gtk\(hyhint
.RE
.BR \-T,
.BR \-\-title
=
.I title
.RS
Use
.I title
in window title (Turns off
.I allowChangingTitle
option.)
.RE
.P
termit \-\-init=session.lua \-\-name=TermitName \-\-class=TermitClass \-\-role=TermitRole \-\-title=TermitTitle \-\-execute=zsh
.P
.RE
.SH FILES
.I $HOME/.config/termit/rc.lua
.RS
Per user configuration file. See section
.BR "TERMIT LUA API"
for further details.
.RE
.I /usr/share/doc/termit/rc.lua.example
.RS
Example rc.lua file. Demonstrates usage of almost all available settings.
.SH "TERMIT LUA API"
.B "Functions:"

.B activateTab
(
.I tab_index
)
    Activates tab by index.
    tab_index \(hy index of tab to activate. Index of the first tab is 1.
.P
.B addMenu
(
.I menu
)
    Adds menu in menubar.
    menu \(hy table of TermitMenuItem type.
.P
.B addPopupMenu
(
.I menu
)
    Adds menu in popup menu, similar to addMenu.
    menu \(hy table of TermitMenuItem type.
.P
.B bindKey
(
.I keys
,
.I luaFunction
)
    Sets new keybinding. If luaFunction is 
.I nil
removes keybinding.
    keys \(hy string with keybinding. Available modifiers are Alt Ctrl Shift Meta Super Hyper.
    luaFunction \(hy callback function
.P
.B bindMouse
(
.I event
,
.I luaFunction
)
    Sets new mouse\(hybinding. If luaFunction is 
.I nil
removes mouse\(hybinding.
    event \(hy string with one of those values: DoubleClick
    luaFunction \(hy callback function
.P
.B closeTab
()
    Closes active tab.
.P
.B copy
()
    Copies selection into tab's buffer.
.P
.B currentTab
()
    Returns current tab of TermitTabInfo.
.P
.B currentTabIndex
()
    Returns current tab index.
.P
.B feed
(
.I
data
)
    Interprets data as if it were data received from a terminal process.
.P
.B feedChild
(
.I
data
)
    Sends a data to the terminal as if it were data entered by the user at the keyboard.
.P
.B findDlg
()
    Opens search entry.
.P
.B findNext
()
    Searches the next string matching search regex.
.P
.B findPrev
()
    Searches the previous string matching search regex.
.P
.B forEachRow
(
.I func
)
    For each terminal row in entire scrollback buffer executes function passing row as argument.
    func \(hy function to be called.
.P
.B forEachVisibleRow
(
.I func
)
    For each visible terminal row executes function passing row as argument.
    func \(hy function to be called.
.P
.B loadSessionDlg
()
    Displays "Load session" dialog.
.P
.B nextTab
()
    Activates next tab.
.P
.B openTab
(
.I tabInfo
)
    Opens new tab.
    tabinfo \(hy table of TermitTabInfo with tab settings.
.P
.B paste
()
    Pastes tab's buffer.
.P
.B preferencesDlg
()
    Displays "Preferences" dialog.
.P
.B prevTab
()
    Activates previous tab.
.P
.B quit
()
    Quit.
.P
.B reconfigure
()
    Sets all configurable variables to defaults and forces rereading rc.lua.
.P
.B saveSessionDlg
()
    Displays "Save session" dialog.
.P
.B selection
()
    Returns selected text from current tab.
.P
.B setColormap
(
.I colormap
)
    Changes colormap for active tab.
    colormap \(hy array with 8 or 16 or 24 colors.
.P
.B setEncoding
(
.I encoding
)
    Changes encoding for active tab.
    encoding \(hy string with encoding name.
.P
.B setKbPolicy
(
.I kb_policy
)
    Sets keyuboard policy for shortcuts.
    kb_policy \(hy string with one of those values:
        keycode \(hy use hardware keyboard codes (XkbLayout\(hyindependent)
        keysym \(hy use keysym values (XkbLayout\(hydependent)
.P
.B setOptions
(
.I opts
)
    Changes default options.
    opts \(hy TermitOptions table with new options.
.P
.B setTabBackgroundColor
(
.I color
)
    Changes background color for active tab.
    color \(hy string with new color.
.P
.B setTabFont
(
.I font
)
    Changes font for active tab.
    font \(hy string with new font.
.P
.B setTabForegroundColor
(
.I color
)
    Changes foreground (e.g. font) color for active tab.
    color \(hy string with new color.
.P
.B setTabPos
(
.I newPos
)
    Changes position for active tab.
    newPos \(hy number with new tab position.
.P
.B setTabTitle
(
.I tabTitle
)
    Changes title for active tab.
    tabTitle \(hy string with new tab title.
.P
.B setTabTitleDlg
()
    Displays "Set tab title" dialog.
.P
.B setWindowTitle
(
.I title
)
    Changes termit window title.
    title \(hy string with new title.
.P
.B spawn
(
.I command
)
    Spawns command (works via shell).
    command \(hy string with command and arguments.
.P
.B toggleMenubar
()
    Displays or hides menubar.
.P
.P
.B toggleTabbar
()
    Displays or hides tabbar.
.P

.B "Types:"

.B TermitCursorBlinkMode
\(hy one of those string values:
    System              Follow GTK+ settings for cursor blinking
    BlinkOn             Cursor blinks
    BlinkOff            Cursor does not blink
.P
.B TermitCursorShape
\(hy one of those string values:
    Block               Draw a block cursor
    Ibeam               Draw a vertical bar on the left side of character
    Underline           Draw a horizontal bar below the character
.P
.B TermitEraseBinding
\(hy one of those string values:
    Auto                VTE_ERASE_AUTO
    AsciiBksp           VTE_ERASE_ASCII_BACKSPACE
    AsciiDel            VTE_ERASE_ASCII_DELETE
    EraseDel            VTE_ERASE_DELETE_SEQUENCE
    EraseTty            VTE_ERASE_TTY
.P
For detailed description look into Vte docs.
.P
.B TermitKeybindings
\(hy table with predefined keybindings.
    closeTab            'Ctrl\(hyw'
    copy                'Ctrl\(hyInsert'
    nextTab             'Alt\(hyRight'
    openTab             'Ctrl\(hyt'
    paste               'Shift\(hyInsert'
    prevTab             'Alt\(hyLeft'
.P
.B TermitMatch
\(hy table for matches.
    field name          match regular expression
    field value         lua callback for action on Left\(hyclick.
.P
.B TermitMenuItem
\(hy table for menuitems.
    accel               accelerator for menuitem. String with keybinding
    action              lua function to execute when item activated
    name                name for menuitem
.P
.B TermitOptions
\(hy table with termit options.
    allowChangingTitle  auto change title (similar to xterm)

ChangeLog
	new options:
		topMenu, tabCloseButton
	removed geometry hints
version 3.1
	require Lua 5.3
	disable menubar on F10
	new options:
		scrollOnKeystroke, scrollOnOutput
	fixed tab style reset after cancel preferences dialog
	refactored several deprecated functions
	fixed recursive termit_close_tab call
	appdata -> metainfo
	default icon name for terminal window
	irc colormap (by clort81)
	get rid of deprecated match_check
version 3.0
	added AppData file
	license changed to GPLv3.0+
	added 'apply to all tabs' in preferences dialog
	new options:
		cursorBlinkMode, cursorShape
		startMaximized, hideTitlebarWhenMaximized
	wordChars replaced with wordCharExceptions
	removed options:
		imageFile, transparency, visibleBell
	Require libvte>0.29 and Gtk-3.
version 2.10.1
	Fixed error with conflicting libvte/Gtk versions
	Fixed ebuild (by Ilya Mochalov).
	Suppress "invalid option" for -e.
	Fixed missing link to math library (by Thomas Koch)
	Miminum required version for Gtk-2 changed to 2.24.
	Miminum required version for libvte changed to 0.28.
	Rebuild po-files due to source changes.
version 2.10.0
	fixed compilation for OpenBSD
	prevent building termit with incompatible Gtk and Vte versions
	"exec" commandline option now works nearly like in xterm
	new options:
		showBorder - show notebook's borders
version 2.9.6
	fixed handling geometry setting
	added support to choose libraries while configuring termit build
	scrollable tabs in tabbar
	new options:
		tabPos - set tabbar position
version 2.9.5
	fixed bug with hideTabbar and hideSingleTab
	tabs indexes start with 1 (Lua-indexing)
	fixed desktop file
	added Swedish translation (by Tomas Fasth)
version 2.9.4
	fix hyphen-used-as-minus-sign in manpage (by Thomas Koch)
	new options:
		hideTabbar (by Anurag Priyam)
	new Lua functions:
		toggleTabbar
	added Polish translation (by Miś Uszatek)
	added Danish translation
	added support for Lua 5.2
	fixed docs
version 2.9.3
	fixed installing as non-root
	added Spanish translation (by Martintxo)
	added German translation (by Dennis Ploeger)
	updated Hungarian translation (by Peter Polonkai)
	apply foreground color for bold fonts
	added several WM-specific commandline options: --name, --class, --role
version 2.9.2
	fix build with --as-needed linker flags enabled (patch by Artur Frysiak)
	Meta, Super and Hyper in keybindings (patch by Artur Frysiak)
	project language set to C
	new Lua functions:
	    feed feedChild
	fixed some compilation issues when using Gtk+3
	added Japanese translation (by Nobuhiro Iwamatsu)
version 2.9.1
	fixed setEncoding in termit.utils module
	fixed possible zombie process
version 2.9.0
	new Lua functions:
		setColormap
	added support for system-wide rc.lua
	added support for using modules in rc.lua
	added several predefined colormaps
version 2.8.1
	fixed bug in currentTabIndex
version 2.8.0
	search in vte buffer (only for vte >= 0.26)
	new Lua functions
		forEachRow
		forEachVisibleRow
		findDlg findNext findPrev
	new options
		backspaceBinding
		deleteBinding
		setStatusbar
	extra warnings when compiling in DEBUG mode
	fixed docs
version 2.7.1
	display proper encoding when switching pages
version 2.7.0
	refactored reading table from Lua
	fixed working with sessions
	new options:
		imageFile - path to image to be set on the background
		colormap - instead of setColormap
		matches - instead of setMatches
		tabs - opened by default tabs
	changed default binding for 'closeTab' from 'Ctrl-w' to 'CtrlShift-w'

src/termit.c
/*  Copyright (C) 2007-2010, Evgeny Ratnikov
    Copyright (C) 2018, Roberto Vergaray

    This file is part of hermit, forked from termit.
    hermit is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 2
    as published by the Free Software Foundation.
    hermit is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    You should have received a copy of the GNU General Public License
    along with hermit. If not, see <http://www.gnu.org/licenses/>.*/

#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <gtk/gtk.h>
#include <gdk/gdkkeysyms.h>
#include <vte/vte.h>

#include "config.h"

#include "termit_core_api.h"
#include "lua_api.h"
#include "callbacks.h"
#include "sessions.h"
#include "keybindings.h"
#include "configs.h"
#include "termit.h"

struct HermitData hermit = {};

struct HermitTab* hermit_get_tab_by_index(guint index)
{
    GtkWidget* tabWidget = gtk_notebook_get_nth_page(GTK_NOTEBOOK(hermit.notebook), index);
    if (!tabWidget) {
        ERROR("tabWidget at %u is NULL", index);
        return NULL;
    }
    struct HermitTab* pTab = (struct HermitTab*)g_object_get_data(G_OBJECT(tabWidget), HERMIT_TAB_DATA);
    return pTab;
}

static GtkWidget* create_statusbar()
{
    GtkWidget* statusbar = gtk_statusbar_new();
    return statusbar;
}

static void create_search(struct HermitData* hermit)
{
    hermit->b_toggle_search = gtk_toggle_button_new();
    gtk_button_set_image(GTK_BUTTON(hermit->b_toggle_search),
            gtk_image_new_from_stock(GTK_STOCK_FIND, GTK_ICON_SIZE_BUTTON));
    g_signal_connect(G_OBJECT(hermit->b_toggle_search), "toggled", G_CALLBACK(hermit_on_toggle_search), NULL);

    hermit->b_find_next = gtk_button_new();
    gtk_button_set_image(GTK_BUTTON(hermit->b_find_next),
            gtk_image_new_from_stock(GTK_STOCK_GO_FORWARD, GTK_ICON_SIZE_BUTTON));
    g_signal_connect(G_OBJECT(hermit->b_find_next), "clicked", G_CALLBACK(hermit_on_find_next), NULL);

    hermit->b_find_prev = gtk_button_new();
    gtk_button_set_image(GTK_BUTTON(hermit->b_find_prev),
            gtk_image_new_from_stock(GTK_STOCK_GO_BACK, GTK_ICON_SIZE_BUTTON));
    g_signal_connect(G_OBJECT(hermit->b_find_prev), "clicked", G_CALLBACK(hermit_on_find_prev), NULL);

    hermit->search_entry = gtk_entry_new();
    g_signal_connect(G_OBJECT(hermit->search_entry), "key-press-event", G_CALLBACK(hermit_on_search_keypress), NULL);
}

static void pack_widgets()
{
    hermit.vbox = gtk_vbox_new(FALSE, 0);
    hermit.hbox = gtk_hbox_new(FALSE, 0);
    gtk_box_pack_start((configs.top_menu)?GTK_BOX(hermit.vbox):GTK_BOX(hermit.hbox), hermit.menu_bar, FALSE, 0, 0);

    gtk_box_pack_start(GTK_BOX(hermit.hbox), hermit.b_toggle_search, FALSE, 0, 0);
    gtk_box_pack_start(GTK_BOX(hermit.hbox), hermit.search_entry, FALSE, 0, 0);
    gtk_box_pack_start(GTK_BOX(hermit.hbox), hermit.b_find_prev, FALSE, 0, 0);
    gtk_box_pack_start(GTK_BOX(hermit.hbox), hermit.b_find_next, FALSE, 0, 0);
    gtk_box_pack_start(GTK_BOX(hermit.hbox), hermit.statusbar, TRUE, 1, 0);

    gtk_box_pack_start(GTK_BOX(hermit.vbox), hermit.notebook, TRUE, 1, 0);
    gtk_box_pack_start(GTK_BOX(hermit.vbox), hermit.hbox, FALSE, 1, 0);
    gtk_container_add(GTK_CONTAINER(hermit.main_window), hermit.vbox);

    if (!gtk_notebook_get_n_pages(GTK_NOTEBOOK(hermit.notebook)))
        hermit_append_tab();
    g_signal_connect(G_OBJECT(hermit.notebook), "switch-page", G_CALLBACK(hermit_on_switch_page), NULL);
    g_signal_connect(G_OBJECT(hermit.main_window), "button-press-event", G_CALLBACK(hermit_on_double_click), NULL);
}

static void menu_item_set_accel(GtkMenuItem* mi, const gchar* parentName,
        const gchar* itemName, const gchar* accel)
{
    struct KeyWithState kws = {};
    if (hermit_parse_keys_str(accel, &kws) < 0) {
        ERROR("failed to parse keybinding: %s", accel);
        return;
    }
    gchar* path = NULL;
    path = g_strdup_printf("<Hermit>/%s/%s", parentName, itemName);
    gtk_menu_item_set_accel_path(GTK_MENU_ITEM(mi), path);
    gtk_accel_map_add_entry(path, kws.keyval, kws.state);
    g_free(path);
}

static void hermit_create_menus(GtkWidget* menu_bar, GtkAccelGroup* accel, GArray* menus)
{
    TRACE("menus->len=%d", menus->len);
    guint j = 0;
    for (; j<menus->len; ++j) {
        struct UserMenu* um = &g_array_index(menus, struct UserMenu, j);

        GtkWidget *mi_menu = gtk_menu_item_new_with_label(um->name);
        GtkWidget *menu = gtk_menu_new();

        gtk_menu_item_set_submenu(GTK_MENU_ITEM(mi_menu), menu);
        gtk_menu_set_accel_group(GTK_MENU(menu), accel);

        TRACE("%s items->len=%d", um->name, um->items->len);
        guint i = 0;
        gtk_menu_set_accel_group(GTK_MENU(menu), accel);
        for (; i<um->items->len; ++i) {
            struct UserMenuItem* umi = &g_array_index(um->items, struct UserMenuItem, i);
            GtkWidget *mi_tmp = gtk_menu_item_new_with_label(umi->name);
            g_object_set_data(G_OBJECT(mi_tmp), HERMIT_USER_MENU_ITEM_DATA, umi);
            g_signal_connect(G_OBJECT(mi_tmp), "activate", G_CALLBACK(hermit_on_menu_item_selected), NULL);
            if (umi->accel) {
                menu_item_set_accel(GTK_MENU_ITEM(mi_tmp), um->name, umi->name, umi->accel);
            }
            gtk_menu_shell_append(GTK_MENU_SHELL(menu), mi_tmp);
        }
        gtk_menu_shell_append(GTK_MENU_SHELL(menu_bar), mi_menu);
    }
}

static GtkWidget* hermit_lua_menu_item_from_stock(const gchar* stock_id, const char* luaFunc)
{
    GtkWidget *mi = gtk_image_menu_item_new_from_stock(stock_id, NULL);
    g_signal_connect(G_OBJECT(mi), "activate", G_CALLBACK(hermit_on_menu_item_selected), NULL);
    struct UserMenuItem* umi = (struct UserMenuItem*)malloc(sizeof(struct UserMenuItem));
    umi->name = g_strdup(gtk_menu_item_get_label(GTK_MENU_ITEM(mi)));
    umi->accel = NULL;
    umi->lua_callback = hermit_get_lua_func(luaFunc);
    g_object_set_data(G_OBJECT(mi), HERMIT_USER_MENU_ITEM_DATA, umi);
    return mi;
}

static GtkWidget* hermit_lua_menu_item_from_string(const gchar* label, const char* luaFunc)
{
    GtkWidget *mi = gtk_menu_item_new_with_label(label);
    g_signal_connect(G_OBJECT(mi), "activate", G_CALLBACK(hermit_on_menu_item_selected), NULL);
    struct UserMenuItem* umi = (struct UserMenuItem*)malloc(sizeof(struct UserMenuItem));
    umi->name = g_strdup(gtk_menu_item_get_label(GTK_MENU_ITEM(mi)));
    umi->accel = NULL;
    umi->lua_callback = hermit_get_lua_func(luaFunc);
    g_object_set_data(G_OBJECT(mi), HERMIT_USER_MENU_ITEM_DATA, umi);
    return mi;
}

void hermit_create_menubar()
{

    GtkAccelGroup *accel = gtk_accel_group_new();
    gtk_window_add_accel_group(GTK_WINDOW(hermit.main_window), accel);
    GtkWidget* menu_bar = gtk_menu_bar_new();

    // File menu
    GtkWidget *mi_new_tab = hermit_lua_menu_item_from_stock(GTK_STOCK_ADD, "openTab");
    GtkWidget *mi_close_tab = hermit_lua_menu_item_from_stock(GTK_STOCK_DELETE, "closeTab");
    GtkWidget *mi_exit = hermit_lua_menu_item_from_stock(GTK_STOCK_QUIT, "quit");

    GtkWidget *mi_file = gtk_menu_item_new_with_label(_("File"));
    GtkWidget *file_menu = gtk_menu_new();

    gtk_menu_shell_append(GTK_MENU_SHELL(file_menu), mi_new_tab);
    gtk_menu_shell_append(GTK_MENU_SHELL(file_menu), mi_close_tab);
    gtk_menu_shell_append(GTK_MENU_SHELL(file_menu), gtk_separator_menu_item_new());
    gtk_menu_shell_append(GTK_MENU_SHELL(file_menu), mi_exit);
    gtk_menu_item_set_submenu(GTK_MENU_ITEM(mi_file), file_menu);

    gtk_menu_shell_append(GTK_MENU_SHELL(menu_bar), mi_file);

    // Edit menu
    GtkWidget *mi_set_tab_name = hermit_lua_menu_item_from_string(_("Set tab name..."), "setTabTitleDlg");
    GtkWidget *mi_edit_preferences = hermit_lua_menu_item_from_stock(GTK_STOCK_PREFERENCES, "preferencesDlg");
    GtkWidget *mi_copy = hermit_lua_menu_item_from_stock(GTK_STOCK_COPY, "copy");
    GtkWidget *mi_paste = hermit_lua_menu_item_from_stock(GTK_STOCK_PASTE, "paste");

    GtkWidget *mi_edit = gtk_menu_item_new_with_label(_("Edit"));
    GtkWidget *edit_menu = gtk_menu_new();
    gtk_menu_shell_append(GTK_MENU_SHELL(edit_menu), mi_copy);
    gtk_menu_shell_append(GTK_MENU_SHELL(edit_menu), mi_paste);
    gtk_menu_shell_append(GTK_MENU_SHELL(edit_menu), gtk_separator_menu_item_new());
    gtk_menu_shell_append(GTK_MENU_SHELL(edit_menu), mi_set_tab_name);
    gtk_menu_shell_append(GTK_MENU_SHELL(edit_menu), mi_edit_preferences);
    gtk_menu_item_set_submenu(GTK_MENU_ITEM(mi_edit), edit_menu);

    gtk_menu_shell_append(GTK_MENU_SHELL(menu_bar), mi_edit);

    // Sessions menu
    GtkWidget *mi_load_session = hermit_lua_menu_item_from_stock(GTK_STOCK_OPEN, "loadSessionDlg");
    GtkWidget *mi_save_session = hermit_lua_menu_item_from_stock(GTK_STOCK_SAVE, "saveSessionDlg");

    GtkWidget *mi_sessions = gtk_menu_item_new_with_label(_("Sessions"));
    GtkWidget *sessions_menu = gtk_menu_new();
    gtk_menu_shell_append(GTK_MENU_SHELL(sessions_menu), mi_load_session);
    gtk_menu_shell_append(GTK_MENU_SHELL(sessions_menu), mi_save_session);
    gtk_menu_item_set_submenu(GTK_MENU_ITEM(mi_sessions), sessions_menu);

    gtk_menu_shell_append(GTK_MENU_SHELL(menu_bar), mi_sessions);

    // User menus
    hermit_create_menus(menu_bar, accel, configs.user_menus);

    hermit.menu_bar = menu_bar;
}

void hermit_create_popup_menu()
{
    hermit.menu = gtk_menu_new();

    GtkWidget *mi_new_tab = hermit_lua_menu_item_from_stock(GTK_STOCK_ADD, "openTab");
    GtkWidget *mi_close_tab = hermit_lua_menu_item_from_stock(GTK_STOCK_DELETE, "closeTab");
    GtkWidget *mi_set_tab_name = hermit_lua_menu_item_from_string(_("Set tab name..."), "setTabTitleDlg");
    GtkWidget *mi_edit_preferences = hermit_lua_menu_item_from_stock(GTK_STOCK_PREFERENCES, "preferencesDlg");
    GtkWidget *mi_copy = hermit_lua_menu_item_from_stock(GTK_STOCK_COPY, "copy");
    GtkWidget *mi_paste = hermit_lua_menu_item_from_stock(GTK_STOCK_PASTE, "paste");
    GtkWidget *mi_exit = hermit_lua_menu_item_from_stock(GTK_STOCK_QUIT, "quit");
    hermit.mi_show_scrollbar = gtk_check_menu_item_new_with_label(_("Scrollbar"));
    gtk_check_menu_item_set_active(GTK_CHECK_MENU_ITEM(hermit.mi_show_scrollbar), configs.show_scrollbar);
    hermit_set_show_scrollbar_signal(hermit.mi_show_scrollbar, NULL);

    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), mi_new_tab);
    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), mi_close_tab);
    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), gtk_separator_menu_item_new());
    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), mi_set_tab_name);
    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), hermit.mi_show_scrollbar);
    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), mi_edit_preferences);
    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), gtk_separator_menu_item_new());
    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), mi_copy);
    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), mi_paste);
    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), gtk_separator_menu_item_new());
    gtk_menu_shell_append(GTK_MENU_SHELL(hermit.menu), mi_exit);

    // User popup menus
    TRACE("user_popup_menus->len=%d", configs.user_popup_menus->len);
    guint j = 0;
    for (; j<configs.user_popup_menus->len; ++j) {
        struct UserMenu* um = &g_array_index(configs.user_popup_menus, struct UserMenu, j);

        GtkWidget *mi_util = gtk_menu_item_new_with_label(um->name);
        GtkWidget *utils_menu = gtk_menu_new();
        gtk_menu_item_set_submenu(GTK_MENU_ITEM(mi_util), utils_menu);

        TRACE("%s items->len=%d", um->name, um->items->len);
        guint i = 0;
        for (; i<um->items->len; i++) {
            struct UserMenuItem* umi = &g_array_index(um->items, struct UserMenuItem, i);
            GtkWidget *mi_tmp = gtk_menu_item_new_with_label(umi->name);
            g_object_set_data(G_OBJECT(mi_tmp), HERMIT_USER_MENU_ITEM_DATA, umi);
            gtk_menu_shell_append(GTK_MENU_SHELL(utils_menu), mi_tmp);
            g_signal_connect(G_OBJECT(mi_tmp), "activate",
                G_CALLBACK(hermit_on_menu_item_selected), NULL);
        }
        gtk_menu_shell_insert(GTK_MENU_SHELL(hermit.menu), mi_util, 6);
    }
    gtk_widget_show_all(hermit.menu);
}

static void hermit_init(const gchar* initFile, gchar** argv)
{
    hermit_init_sessions();
    hermit_configs_set_defaults();

    hermit.tab_max_number = 1;

    hermit.main_window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
    hermit.statusbar = create_statusbar();
    create_search(&hermit);
    hermit.notebook = gtk_notebook_new();
    gtk_notebook_set_scrollable(GTK_NOTEBOOK(hermit.notebook), TRUE);
    gtk_notebook_set_show_tabs(GTK_NOTEBOOK(hermit.notebook), TRUE);

    hermit_lua_init(initFile);

    gtk_notebook_set_tab_pos(GTK_NOTEBOOK(hermit.notebook), configs.tab_pos);

    if (argv) {
        hermit_append_tab_with_command(argv);
    }

    hermit_create_menubar();
    pack_widgets();
    hermit_create_popup_menu();

    if (!configs.allow_changing_title) {
        hermit_set_window_title(configs.default_window_title);
    }
    gtk_notebook_set_show_border(GTK_NOTEBOOK(hermit.notebook), configs.show_border);
}

enum {
    HERMIT_GETOPT_HELP = 'h',
    HERMIT_GETOPT_VERSION = 'v',
    HERMIT_GETOPT_INIT = 'i',
    HERMIT_GETOPT_NAME = 'n',
    HERMIT_GETOPT_CLASS = 'c',
    HERMIT_GETOPT_ROLE = 'r',
    HERMIT_GETOPT_TITLE = 'T'
};

static void hermit_print_usage()
{
    g_print(
"hermit %s - terminal emulator\n"
"\n"
"This program comes with NO WARRANTY, to the extent permitted by law.\n"
"You may redistribute copies of this program\n"
"under the terms of the GNU General Public License.\n"
"For more information about these matters, see the file named COPYING.\n"
"\n"
"Options:\n"
"  -h, --help             - print this help message\n"
"  -v, --version          - print version number\n"
"  -e, --execute          - execute command\n"
"  -i, --init=init_file   - use init_file instead of standard rc.lua\n"
"  -n, --name=name        - set window name hint\n"
"  -c, --class=class      - set window class hint\n"
"  -r, --role=role        - set window role (Gtk hint)\n"
"  -T, --title=title      - set window title\n"
"", PACKAGE_VERSION);
}
/*
    Support the command-line option "-e <command>", which creates a new
    terminal window and runs the specified command.  <command> may be
    multiple arguments, which form the argument list to the executed
    program.  In other words, the behavior is as though the arguments
    were passed directly to execvp, bypassing the shell.  (xterm's
    behavior of falling back on using the shell if -e had a single
    argument and exec failed is permissible but not required.)

    exec($T, "-e", "bc")              --> runs bc
    exec($T, "-e", "bc --quiet")      --> fails OR runs bc --quiet [1]
    exec($T, "-e", "bc", "--quiet")   --> runs bc --quiet
    exec($T, "-e", "bc *")            --> fails OR runs bc * (wildcard) [1]
    exec($T, "-e", "bc", "*")         --> runs bc * (literal)

[1] Programs are allowed to pass the entire argument to -e to the shell
    if and only if there is only one argument and exec fails.
*/

static GArray* parse_execute_args(int argc, char **argv)
{
    GArray* arr = NULL;
    int i = 0;
    int foundE = 0;
    for (; i < argc; ++i) {
        if (foundE) {
            gchar* arg = g_strdup(argv[i]);
            g_array_append_val(arr, arg);
        } else {
            TRACE("[%s] %d", argv[i], strncmp(argv[i], "--execute=", 10));
            if ((strcmp(argv[i], "-e") == 0)
                    || (strcmp(argv[i], "--execute") == 0)) {
                foundE = 1;
                arr = g_array_new(FALSE, TRUE, sizeof(gchar*));
            } else if (strncmp(argv[i], "--execute=", 10) == 0) {
                foundE = 1;
                TRACE("arg=[%s]", argv[i] + 10);
                arr = g_array_new(FALSE, TRUE, sizeof(gchar*));
                gchar* arg = g_strdup(argv[i] + 10);
                g_array_append_val(arr, arg);
            }
        }
    }
    if (arr == NULL) {
        return NULL;
    }
    if (arr->len > 0) {
        return arr;
    }
    g_array_free(arr, TRUE);
    return NULL;
}

int main(int argc, char **argv)
{
    opterr = 0; // suppress "invalid option" error message from getopt
    gchar* initFile = NULL;
    GArray* arrArgv = parse_execute_args(argc, argv);;
    gchar *windowName = NULL, *windowClass = NULL, *windowRole = NULL, *windowTitle = NULL;
    while (1) {
        static struct option long_options[] = {
            {"help", no_argument, 0, HERMIT_GETOPT_HELP},
            {"version", no_argument, 0, HERMIT_GETOPT_VERSION},
            {"init", required_argument, 0, HERMIT_GETOPT_INIT},
            {"name", required_argument, 0, HERMIT_GETOPT_NAME},
            {"class", required_argument, 0, HERMIT_GETOPT_CLASS},
            {"role", required_argument, 0, HERMIT_GETOPT_ROLE},
            {"title", required_argument, 0, HERMIT_GETOPT_TITLE},
            {0, 0, 0, 0}
        };
        /* getopt_long stores the option index here. */
        int option_index = 0;

        int flag = getopt_long(argc, argv, "-hvi:n:c:r:T:", long_options, &option_index);

        /* Detect the end of the options. */
        if (flag == -1)
            break;

        switch (flag) {
        case HERMIT_GETOPT_HELP:
            hermit_print_usage();
            return 0;
        case HERMIT_GETOPT_VERSION:
            g_printf(PACKAGE_VERSION);
            g_printf("\n");
            return 0;
        case HERMIT_GETOPT_INIT:
            initFile = g_strdup(optarg);
            break;
        case HERMIT_GETOPT_NAME:
            windowName = g_strdup(optarg);
            break;
        case HERMIT_GETOPT_CLASS:
            windowClass = g_strdup(optarg);
            break;
        case HERMIT_GETOPT_ROLE:
            windowRole = g_strdup(optarg);
            break;
        case HERMIT_GETOPT_TITLE:
            windowTitle = g_strdup(optarg);
            break;
        case '?':
            break;
        /* getopt_long already printed an error message. */
        default:
            break;
        }
    }

    bind_textdomain_codeset(PACKAGE, "UTF-8");
    bindtextdomain(PACKAGE, LOCALEDIR);
    textdomain(PACKAGE);

    gtk_init(&argc, &argv);
#ifdef __linux__
    signal(SIGCHLD, SIG_IGN);
#endif /* LINUX */
    gchar** cmdArgv = NULL;
    if (arrArgv != NULL) {
        cmdArgv = (gchar**)g_new0(gchar*, arrArgv->len + 1);
        guint i = 0;
        for (; i < arrArgv->len; ++i) {
            cmdArgv[i] = g_array_index(arrArgv, gchar*, i);
            TRACE("    %d=[%s]", i, cmdArgv[i]);
        }
        g_array_free(arrArgv, TRUE);
    }
    hermit_init(initFile, cmdArgv);
    g_strfreev(cmdArgv);
    g_free(initFile);

#if 0
    /**
     * dirty hack from gnome-terminal ;-)
     * F10 is used in many console apps, so we change global Gtk setting for hermit
     * */
    gtk_settings_set_string_property(gtk_settings_get_default(), "gtk-menu-bar-accel",
        "<Shift><Control><Mod1><Mod2><Mod3><Mod4><Mod5>F10", "hermit");
#endif // FIXME
    g_signal_connect(G_OBJECT (hermit.main_window), "delete_event", G_CALLBACK (hermit_on_delete_event), NULL);
    g_signal_connect(G_OBJECT (hermit.main_window), "destroy", G_CALLBACK (hermit_on_destroy), NULL);
    g_signal_connect(G_OBJECT (hermit.main_window), "key-press-event", G_CALLBACK(hermit_on_key_press), NULL);

    TRACE("window: came=[%s] class=[%s] role=[%s] title=[%s]",
            windowName, windowClass, windowRole, windowTitle);
    if (windowName || windowClass) {
        gtk_window_set_wmclass(GTK_WINDOW(hermit.main_window), windowName, windowClass);
        g_free(windowName);
        g_free(windowClass);
    }
    if (windowRole) {
        gtk_window_set_role(GTK_WINDOW(hermit.main_window), windowRole);
        g_free(windowRole);
    }
    if (windowTitle) {
        configs.allow_changing_title = FALSE;
        gtk_window_set_title(GTK_WINDOW(hermit.main_window), windowTitle);
        g_free(windowTitle);
    }

    /* Show the application window */
    gtk_widget_show_all(hermit.main_window);

    // actions after display
    hermit_after_show_all();

    gtk_main();
    hermit_config_deinit();
    hermit_lua_close();
    return 0;
}

